File: 백준/Silver/5212. 지구 온난화/지구 온난화.py
import sys
input = sys.stdin.readline

# 입력 받기
R, C = map(int, input().split())
grid = [list(input().strip()) for _ in range(R)]

# 상하좌우 탐색 방향
directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]

# 바다로 변할 위치를 저장할 리스트
to_change = []

# 바다로 변할 섬 확인
for r in range(R):
    for c in range(C):
        if grid[r][c] == 'X':
            sea_count = 0
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                if nr < 0 or nr >= R or nc < 0 or nc >= C or grid[nr][nc] == '.':
                    sea_count += 1
            if sea_count >= 3:
                to_change.append((r, c))

# 섬을 바다로 변환
for r, c in to_change:
    grid[r][c] = '.'

# 남아있는 섬의 영역 확인
min_row, max_row = R, 0
min_col, max_col = C, 0

for r in range(R):
    for c in range(C):
        if grid[r][c] == 'X':
            min_row = min(min_row, r)
            max_row = max(max_row, r)
            min_col = min(min_col, c)
            max_col = max(max_col, c)

# 남아있는 영역 출력
for r in range(min_row, max_row + 1):
    print("".join(grid[r][min_col:max_col + 1]))

File: 백준/Silver/5212. 지구 온난화/지구 온난화.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] input = br.readLine().split(" ");
        int R = Integer.parseInt(input[0]);
        int C = Integer.parseInt(input[1]);

        char[][] grid = new char[R][C];
        for (int r = 0; r < R; r++) {
            String line = br.readLine();
            for (int c = 0; c < C; c++) {
                grid[r][c] = line.charAt(c);
            }
        }

        int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        List<int[]> toChange = new ArrayList<>();

        for (int r = 0; r < R; r++) {
            for (int c = 0; c < C; c++) {
                if (grid[r][c] == 'X') {
                    int seaCount = 0;
                    for (int[] dir : directions) {
                        int nr = r + dir[0];
                        int nc = c + dir[1];
                        if (nr < 0 || nr >= R || nc < 0 || nc >= C || grid[nr][nc] == '.') {
                            seaCount++;
                        }
                    }
                    if (seaCount >= 3) {
                        toChange.add(new int[]{r, c});
                    }
                }
            }
        }

        for (int[] pos : toChange) {
            grid[pos[0]][pos[1]] = '.';
        }

        int minRow = R, maxRow = 0;
        int minCol = C, maxCol = 0;
        
        for (int r = 0; r < R; r++) {
            for (int c = 0; c < C; c++) {
                if (grid[r][c] == 'X') {
                    minRow = Math.min(minRow, r);
                    maxRow = Math.max(maxRow, r);
                    minCol = Math.min(minCol, c);
                    maxCol = Math.max(maxCol, c);
                }
            }
        }

        StringBuilder sb = new StringBuilder();
        if (minRow <= maxRow && minCol <= maxCol) { 
            for (int r = minRow; r <= maxRow; r++) {
                for (int c = minCol; c <= maxCol; c++) {
                    sb.append(grid[r][c]);
                }
                sb.append("\n");
            }
        }
        
        System.out.print(sb.toString());
    }
}